tls/extensions/s2n_client_key_share.c

```diff
@@ -85,19 +83,69 @@ static int s2n_ecdhe_supported_curves_send(struct s2n_connection *conn, struct s
     }
 
     for (size_t i = 0; i < ecc_pref->count; i++) {
-        ecc_evp_params = &conn->secure.client_ecc_evp_params[i];
-        named_curve = ecc_pref->ecc_curves[i];
-        ecc_evp_params->negotiated_curve = named_curve;
-        ecc_evp_params->evp_pkey = NULL;
         /* If a bit in the bitmap (minus the lsb) is set, generate keyshare for the corresponding curve */
         if (S2N_IS_KEY_SHARE_REQUESTED(preferred_key_shares, i)) {
+            ecc_evp_params = &conn->secure.client_ecc_evp_params[i];
+            ecc_evp_params->negotiated_curve = ecc_pref->ecc_curves[i];
+            ecc_evp_params->evp_pkey = NULL;
             GUARD(s2n_ecdhe_parameters_send(ecc_evp_params, out));
         }
     }
 
     return S2N_SUCCESS;
 }
 
+static int s2n_send_hrr_keyshare(struct s2n_connection *conn, struct s2n_stuffer *out)
+{
+    notnull_check(conn);
+    const struct s2n_ecc_named_curve *server_negotiated_curve = NULL;
+    struct s2n_ecc_evp_params *ecc_evp_params = NULL;
+
+    const struct s2n_ecc_preferences *ecc_pref = NULL;
+    GUARD(s2n_connection_get_ecc_preferences(conn, &ecc_pref));
+    notnull_check(ecc_pref);
+
+    server_negotiated_curve = conn->secure.server_ecc_evp_params.negotiated_curve;
+    ENSURE_POSIX(server_negotiated_curve != NULL, S2N_ERR_BAD_KEY_SHARE);
+
+    bool has_supported_curve = false;
+
+    for (size_t i = 0; i < ecc_pref->count; i++) {
+        if (ecc_pref->ecc_curves[i]->iana_id == server_negotiated_curve->iana_id) {
+            has_supported_curve = true;
+            ecc_evp_params = &conn->secure.client_ecc_evp_params[i];
+            GUARD(s2n_ecc_evp_params_free(ecc_evp_params));
```

Comment:
If you free in both branches, it can be moved outside of the if/else.

Reply:
Actually this should be 

```
ENSURE_POSIX(ecc_evp_params == NULL, S2N_ERR_INVALID_HELLO_RETRY);
```

If the keyshare already exists for the server negotiated curve, we are in an invalid HRR path.